### PR TITLE
fix: concurrency conflict in MemPool.TryRemoveUnVerified

### LINKED ISSUE #3499: [`Bug`]: concurrency conflict in `MemPool.TryRemoveUnVerified`
**Describe the bug**

There two write operations in `MemPool.TryRemoveUnVerified`, and this function called by `Blockchain.OnFillMemoryPool` and `MemPool.UpdatePoolForBlockPersisted`.

Write operations triggered by `MemPool.UpdatePoolForBlockPersisted` is protected by `_txRwLock`, but by `Blockchain.OnFillMemoryPool` is not.

```csharp
        internal bool TryRemoveUnVerified(UInt256 hash, [MaybeNullWhen(false)] out PoolItem? item)
        {
            if (!_unverifiedTransactions.TryGetValue(hash, out item))
                return false;

            _unverifiedTransactions.Remove(hash);
            _unverifiedSortedTransactions.Remove(item);
            return true;
        }
```

According to comments of `_txRwLock`, write operations should acquire write lock of `_txRwLock`.

### PR DESCRIPTION
**Pull request by nan01ab**
# Description

Fixes #3499 

## Type of change

- [ ] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/Neo/Ledger/MemoryPool.cs [hunk 1]
```diff
@@ -445,12 +445,19 @@ private void RemoveConflictsOfVerified(PoolItem item)
...
-            return true;
+            _txRwLock.EnterWriteLock();
+            try
+            {
+                if (!_unverifiedTransactions.TryGetValue(hash, out item))
+                    return false;
+        
+                _unverifiedTransactions.Remove(hash);
+                _unverifiedSortedTransactions.Remove(item);
+            }
```
Comment (shargon) [inline src/Neo/Ledger/MemoryPool.cs:456]:
```suggestion
            return true;
            }
```

File: src/Neo/Ledger/MemoryPool.cs [hunk 2]
```diff
@@ -481,7 +489,7 @@ internal void UpdatePoolForBlockPersisted(Block block, DataCache snapshot)
                 // No need to modify VerificationContext as it will be reset afterwards.
                 foreach (Transaction tx in block.Transactions)
                 {
-                    if (!TryRemoveVerified(tx.Hash, out _)) TryRemoveUnVerified(tx.Hash, out _);
+                    if (!TryRemoveVerified(tx.Hash, out _)) TryRemoveUnVerifiedLocked(tx.Hash, out _);
```
Comment (shargon) [inline src/Neo/Ledger/MemoryPool.cs:492]:
```suggestion
                    if (!TryRemoveVerified(tx.Hash, out _)) TryRemoveUnVerified(tx.Hash, out _);
```

File: src/Neo/Ledger/MemoryPool.cs [hunk 3]
```diff
@@ -509,7 +517,7 @@ internal void UpdatePoolForBlockPersisted(Block block, DataCache snapshot)
                 }
                 foreach (var h in stale)
                 {
-                    if (!TryRemoveVerified(h, out _)) TryRemoveUnVerified(h, out _);
+                    if (!TryRemoveVerified(h, out _)) TryRemoveUnVerifiedLocked(h, out _);
```
Comment (shargon) [inline src/Neo/Ledger/MemoryPool.cs:520]:
```suggestion
                    if (!TryRemoveVerified(h, out _)) TryRemoveUnVerified(h, out _);
```

**Review by shargon — DISMISSED** [review]
Require testing

**Comment by nan01ab** [comment]
> Require testing

A unit test added

**Comment by shargon** [comment]
> > Require testing
> 
> A unit test added

In real environment, please @superboyiii could you test it?

**Comment by Jim8y** [comment]
> > > Require testing
> > 
> > 
> > A unit test added
> 
> In real environment, please @superboyiii could you test it?

SuperBoy is on travel, wont be back until 26th.

**Comment by superboyiii** [comment]
> In real environment, please @superboyiii could you test it?

Sure.

**Comment by Hecate2** [comment]
I have doubts on this. Neo uses Akka actors to avoid locks. It was said that a single Akka actor processes only 1 message at any time. This means that no multiple instances of an actor is running. And `TryRemoveUnVerified` (along with `HeaderCache.Count` in the following PR https://github.com/neo-project/neo/pull/3501 , and everything in `MemPool`) is used only by the `Blockchain` actor, in a "single-threaded" way.

**Comment by nan01ab** [comment]
> I have doubts on this. Neo uses Akka actors to avoid locks. It was said that a single Akka actor processes only 1 message at any time. This means that no multiple instances of an actor is running. And `TryRemoveUnVerified` (along with `HeaderCache.Count` in the following PR #3501 , and everything in `MemPool`) is used only by the `Blockchain` actor, in a "single-threaded" way.
> 
> At the definition of `_txRwLock` in `MemoryPool.cs` there are the following comments:
> 
> ```
> The data structures are only modified from the `Blockchain` actor; so operations guaranteed to be
> performed by the blockchain actor do not need to acquire the read lock; they only need the write
> lock for write operations.
> ```
> 
> I am even doubting why write locks are needed, and whether the existing locks are necessary.

`MemPool` and `HeaderCache` is written only by one `actor`, but may read by multi.
`Lock` is not for write-write conflict, but for write-read.

File: tests/Neo.UnitTests/Ledger/UT_MemoryPool.cs [hunk 4]
```diff
@@ -747,6 +747,25 @@ public void TestUpdatePoolForBlockPersisted()
             _unit.VerifiedCount.Should().Be(0);
         }
 
+        [TestMethod]
+        public void TestTryRemoveUnVerified()
```
Comment (Jim8y) [inline tests/Neo.UnitTests/Ledger/UT_MemoryPool.cs:751]: how could this test the conflict?

Comment (nan01ab) [reply tests/Neo.UnitTests/Ledger/UT_MemoryPool.cs:751]: Race condition is hard to test.

**Review by superboyiii — APPROVED** [review]
Tested OK.

**Review by Jim8y — APPROVED** [review]

**Review by shargon — APPROVED** [review]

File: tests/Neo.UnitTests/Ledger/UT_MemoryPool.cs [hunk 5]
```diff
@@ -747,6 +747,25 @@ public void TestUpdatePoolForBlockPersisted()
...
+        {
+            AddTransactions(32);
+            _unit.SortedTxCount.Should().Be(32);
+
+            var txs = _unit.GetSortedVerifiedTransactions().ToArray();
+            _unit.InvalidateVerifiedTransactions();
+
+            _unit.SortedTxCount.Should().Be(0);
+
+            foreach (var tx in txs)
```
Comment (vncoelho) [inline tests/Neo.UnitTests/Ledger/UT_MemoryPool.cs:761]: maybe `_unit.UnVerifiedCount.Should().Be(32);` before the loop

**Review by vncoelho — COMMENTED** [review]
good change

**Pull request merged**